src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,180 @@
+            )
+            for seg in copied_line_buffer
+        )
+        is_empty_line = not any(
+            elem.is_code or elem.is_type("placeholder") for elem in copied_line_buffer
+        )
+        line_summary = self.__class__(
+            line_no=line_no,
+            templated_line=self.templated_line,
+            # Using slicing to copy line_buffer here to be py2 compliant
+            line_buffer=copied_line_buffer,
+            indent_buffer=self.indent_buffer,
+            indent_size=self.indent_size,
+            indent_balance=self.indent_balance,
+            anchor_indent_balance=self.anchor_indent_balance,
+            hanging_indent=self.hanger_pos if self.line_indent_stack else None,
+            # Clean indent is true if the line *ends* with an indent
+            # or has an indent in the initial whitespace.
+            clean_indent=self.clean_indent,
+            # Solidify expensive immutable characteristics
+            templated_line_type=_get_template_block_type(
+                copied_line_buffer, templated_file
+            ),
+            is_comment_line=is_comment_line,
+            is_empty_line=is_empty_line,
+        )
+        # Rest this object so it can continue to be used to track current state
+        self.memory_line_reset()
+        return line_summary
+
+    def set_state_as_of_anchor(
+        self,
+        anchor: Optional[BaseSegment],
+        tab_space_size: int,
+    ):
+        """Create a Line state of this line upon reaching the anchor."""
+        self.anchor_indent_balance = self.indent_balance
+        self.indent_size = _indent_size(
+            self.indent_buffer,
+            tab_space_size=tab_space_size,
+        )
+        self.line_anchor = anchor
+
+        return self
+
+
+def _copy_line_summary(line: _LineSummary) -> _LineSummary:
+    new_line = _LineSummary(**line.__dict__.copy())
+    list_keys = [
+        "line_buffer",
+        "indent_buffer",
+        "line_indent_stack",
+    ]
+    for key in list_keys:
+        el = new_line.__getattribute__(key)
+        new_line.__setattr__(key, el[:])
+    return new_line
+
+
+def _is_clean_indent(prev_line_buffer: List[BaseSegment]):
+    """Check the previous line to see if the current state is a clean indent."""
+    # Assume an unclean indent, but if the last line
+    # ended with an indent then we might be ok.
+    # Was there an indent after the last code element of the previous line?
+    for search_elem in reversed(prev_line_buffer):
+        is_meta = search_elem.is_meta
+        if not search_elem.is_code and not is_meta:
+            continue
+        if is_meta and search_elem.indent_val > 0:  # type: ignore
+            return True
+        break
+
+    return False
+
+
+@dataclasses.dataclass
+class _Memory:
+    # problem_lines keeps track of lines with problems so that we
+    # don't compare to them.
+    problem_lines: List[int] = dataclasses.field(default_factory=list)
```

Comment:
I wonder if some of these could be sets, rather than list. Sets are much more efficient for checking whether an item is in it, but I'm not sure they preserve the sequence of items, if that matters.